merkle/smt/nodes.go

```diff
@@ -28,6 +28,39 @@ type Node struct {
 	Hash []byte
 }
 
+// NodesRow contains nodes at the same tree level sorted by ID from left to
+// right. The IDs of the nodes are unique.
+//
+// TODO(pavelkalinnikov): Hide nodes so that only this package can modify them.
+type NodesRow []Node
+
+// NewNodesRow creates a NodesRow from the given list of nodes. The nodes are
+// reordered in-place if not already sorted.
+func NewNodesRow(nodes []Node) (NodesRow, error) {
+	if len(nodes) == 0 {
+		return nodes, nil
+	}
+	if err := Prepare(nodes, nodes[0].ID.BitLen()); err != nil {
+		return nil, err
+	}
+	return nodes, nil
+}
+
+// inSubtree returns whether all the nodes in this row are strictly under the
+// node with the given ID. Panics if the row is empty.
```

Comment:
I wonder if returning true if the row is empty is better than panicking? This is internal though so we can always revisit.

Reply:
Yeah, I did that first, but then I realised that the wrapping code checks emptiness anyway.

Another thought visited me that an empty set has "all" the properties. I.e. `inSubtree` is a predicate "for all nodes n in this row, n is in the subtree with the given ID" which is true for an empty set, however empty set also satisfies "for all nodes n in this row, n is **not** in the subtree with the given ID". Having at least one node in the list is a "proof" that the updates were indeed intended for this subtree, whereas an empty update could be created for any other subtree and passed to this one by accident.

So, given that, I decided to force the caller of this method to at least **think** about the empty case.